### PR TITLE
Add cell tracking application note

### PR DESCRIPTION
**Pull request by quantumjot**
This PR adds an "applications" page describing the use of the new Tracks layer for cell tracking.

### CODE REVIEW

**Comment by GenevieveBuckley** [comment]
Think I saw a tweet from you earlier about the tracks tutorial, the gif there looked gorgeous

**Comment by quantumjot** [comment]
> Think I saw a tweet from you earlier about the tracks tutorial, the gif there looked gorgeous 

Thanks @GenevieveBuckley! I have incorporated a version of it into the application note now.

**Comment by jni** [comment]
This is great! I love the use of `regionprops_table`. 😊

File: applications/cell_tracking.md [hunk 1]
```diff
@@ -0,0 +1,223 @@
...
+    idx : int
+        Index of the image to load.
+    """
+    filename = os.path.join(PATH, '01_GT/TRA', f'man_track{idx:0>3}.tif')
+    return imread(filename)
+
+stack = np.asarray([load_image(i) for i in range(NUM_IMAGES)])
+```
+
+For each image in the time-lapse sequence, we will now extract the unique track label (`track_id`), centroid and timestamp.
```
Comment (kevinyamauchi) [inline applications/cell_tracking.md:47]:
```suggestion
For each image in the time-lapse sequence, we will now extract the unique track label (`track_id`), centroid and timestamp in order to create the track data we will pass to the `Tracks` layer. For more information on the format of the track data, please see the "tracks data" section of the [tracks layer fundamentals tutorial]({{ '/fundamentals/tracks' | relative_url }}).
```

File: applications/cell_tracking.md [hunk 2]
```diff
@@ -0,0 +1,223 @@
...
+
+# create the final data array, track_id, T, Z, Y, X
+data = data_df.loc[:, ['label','frame','centroid-0','centroid-1','centroid-2']]
+data = data.to_numpy()
+```
+
+### calculating the graph using the lineage information
+
+The `Tracks` layer can also be used to visualize a track 'graph' using the additional keyword argument `graph`. The `graph`  represents associations between tracks, by defining the
+mapping between a `track_id` and the parents of the track. This graph can be useful in single cell tracking to understand the lineage of cells over multiple cell division events.
```
Comment (kevinyamauchi) [inline applications/cell_tracking.md:77]:
```suggestion
mapping between a `track_id` and the parents of the track. This graph can be useful in single cell tracking to understand the lineage of cells over multiple cell division events. For more information on the format of the track `graph`, please see the "tracks graph" section of the [tracks layer fundamentals tutorial]({{ '/fundamentals/tracks' | relative_url }}).
```

File: applications/cell_tracking.md [hunk 3]
```diff
@@ -0,0 +1,223 @@
...
+## summary
+In this application note, we have used napari to track and visualize single cells.
+
+## further reading
+
+References for cell tracking challenge:  
++ http://dx.doi.org/10.1093/bioinformatics/btu080  
++ http://dx.doi.org/10.1038/nmeth.4473  
+
+Arboretum plugin for napari:
```
Comment (kevinyamauchi) [inline applications/cell_tracking.md:218]:
```suggestion
For a more advanced example of visualizing cell tracking data with napari, please see the Arboretum plugin for napari:
```

**Review by kevinyamauchi — APPROVED** [review]
I think this looks great. Thank you, @quantumjot! Super cool stuff. I have made some comments (mostly just linking to other content). Please feel free to modify them as you see fit.

Another thing to note is that there are two links to the images, but I am guessing you were just doing that to be able to see them in the preview. Just a quick reminder to delete the redundant one.

File: applications/cell_tracking.md [hunk 4]
```diff
@@ -0,0 +1,223 @@
...
+
+```python
+def regionprops_plus_time(idx):
+    """Return the unique track label, centroid and time for each track vertex.
+
+    Parameters
+    ----------
+    idx : int
+        Index of the image to calculate the centroids and track labels.
+    """
```
Comment (kevinyamauchi) [inline applications/cell_tracking.md:57]:
```suggestion
    Returns
    -------
    data_df : pd.DataFrame
       The dataframe of track data for one time step (specified by idx).
    """
```

File: applications/cell_tracking.md [hunk 5]
```diff
@@ -0,0 +1,223 @@
...
+NUM_IMAGES = 195
+
+def load_image(idx: int):
+    """Load an image from the sequence.
+
+    Parameters
+    ----------
+    idx : int
+        Index of the image to load.
+    """
```
[the remark this answers is not in the document]
Comment (quantumjot) [reply applications/cell_tracking.md:40]: It's almost as if you're trying to get me to do things properly 😛.

**Comment by quantumjot** [comment]
> I am curious about what the others think, but I usually like to have the complete code available somewhere so people can just copy paste the whole thing if they want to reproduce it. I think either somewhere in the page or in a repo somewhere.

I can link to the `btrack` example notebook. Is it worth adding the full code to napari/examples?

**Comment by quantumjot** [comment]
OK - I think this is pretty much ready to go. Thanks for your feedback @GenevieveBuckley @jni @kevinyamauchi !

File: applications/cell_tracking.md [hunk 6]
```diff
@@ -0,0 +1,236 @@
...
+
+```python
+def root(node: int):
+    """ Recursive function to determine the root node of each graph.
+
+    Parameters
+    ----------
+    node : int
+        the track_id of the starting graph node.
+    """
```
Comment (kevinyamauchi) [inline applications/cell_tracking.md:128]:
```suggestion

    Returns
    -------
    root_id : int
       The track_id of the root of the track specified by idx.
    """
```

**Review by kevinyamauchi — APPROVED** [review]
I think this is good to go! Thanks, @quantumjot . I will leave this up in through west coast time to see if @sofroniewn wants to check it out before merging.

**Comment by kevinyamauchi** [comment]
> I can link to the btrack example notebook. Is it worth adding the full code to napari/examples?

I think the link to the notebook that you added is perfect!

File: applications/cell_tracking.md [hunk 7]
```diff
@@ -0,0 +1,241 @@
...
+    idx : int
+        Index of the image to calculate the centroids and track labels.
+
+    Returns
+    -------
+    data_df : pd.DataFrame
+       The dataframe of track data for one time step (specified by idx).
+    """
+    props = regionprops_table(stack[idx, ...], properties=('label', 'centroid'))
+    props['frame'] = np.ones(props['label'].shape) * idx
```
Comment (jni) [inline applications/cell_tracking.md:70]:
```suggestion
    props['frame'] = np.full(props['label'].shape, idx)
```

**Review by jni — APPROVED** [review]
@quantumjot This is great! I've made what I hope is one last set of suggestions! :grimacing: I can push also if you'd like me to take this over!

File: applications/cell_tracking.md [hunk 8]
```diff
@@ -0,0 +1,241 @@
...
+
+```python
+with napari.gui_qt():
+    viewer = napari.Viewer()
+    viewer.add_tracks(data, properties=properties, graph=graph)
+```
+
+![image]({{ '/assets/tutorials/tracks_btrack.png' | relative_url }})
+
+A notebook for this example can be found in the btrack examples directory ([`napari_btrack.ipynb`](https://github.com/quantumjot/BayesianTracker/blob/master/examples/napari_btrack.ipynb))
```
[the remark this answers is not in the document]
Comment (quantumjot) [reply applications/cell_tracking.md:224]: Will do.

File: applications/cell_tracking.md [hunk 9]
```diff
@@ -0,0 +1,241 @@
...
+    tracker.track_interactive(step_size=100)
+    tracker.optimize()
+
+    # get the tracks in a format for napari visualization
+    data, properties, graph = tracker.to_napari(ndim=2)
+```
+
+We set the configuration of the tracker using a configuration file using the `.configure_from_file()` method. An the example configuration file can be found [here](https://github.com/quantumjot/BayesianTracker/blob/master/models/cell_config.json).
+
+The objects loaded earlier are added to the tracker using the `.append()` method. We also set the imaging volume using the `volume` property. As this is a 2D dataset, the limits of the volume are set to the XY dimensions of the image dataset, and the Z dimension is set to be large (±1e5).
```
[the remark this answers is not in the document]
Comment (quantumjot) [reply applications/cell_tracking.md:210]: I'll add a better explanation here - it has to do with the limits of the volume versus the limits of the field of view.

File: applications/cell_tracking.md [hunk 10]
```diff
@@ -0,0 +1,241 @@
...
+
+# sort the data lexicographically by track_id then time
+data_df = data_df.reset_index(drop=True)
+data_df = data_df.sort_values(['label', 'frame'], ignore_index=True)
+
+# create the final data array, track_id, T, Z, Y, X
+data = data_df.loc[:, ['label','frame','centroid-0','centroid-1','centroid-2']]
+data = data.to_numpy()
+```
+
```
[the remark this answers is not in the document]
Comment (quantumjot) [reply applications/cell_tracking.md:84]: Sounds good. I can add this in.

File: applications/cell_tracking.md [hunk 11]
```diff
@@ -0,0 +1,241 @@
...
+To extract the graph, we load the text file and convert it to a Nx4 integer numpy array, where the rows represent individual tracks and the columns represent L, B, E and P:
+```python
+lbep = np.loadtxt(os.path.join(PATH, '01_GT/TRA', 'man_track.txt'), dtype=np.uint)
+```
+
+We can then create a dictionary representing the graph, where the key is the unique track label (L) and the value is the label of the parent track (P). We make a small change to the format here, to set the
+parent label equal to the track label if there is no parent defined.  
+
+```python
+full_graph = {row[0]: row[3] if row[3]>0 else row[0] for row in lbep}
```
[the remark this answers is not in the document]
Comment (quantumjot) [reply applications/cell_tracking.md:108]: I agree - this is nice!

File: applications/cell_tracking.md [hunk 12]
```diff
@@ -0,0 +1,241 @@
...
+> E - a zero-based temporal index of the frame in which the track ends  
+> P - label of the parent track (0 is used when no parent is defined)  
+
+To extract the graph, we load the text file and convert it to a Nx4 integer numpy array, where the rows represent individual tracks and the columns represent L, B, E and P:
+```python
+lbep = np.loadtxt(os.path.join(PATH, '01_GT/TRA', 'man_track.txt'), dtype=np.uint)
+```
+
+We can then create a dictionary representing the graph, where the key is the unique track label (L) and the value is the label of the parent track (P). We make a small change to the format here, to set the
+parent label equal to the track label if there is no parent defined.
```
[the remark this answers is not in the document]
Comment (quantumjot) [reply applications/cell_tracking.md:105]: A good question, and one that I do not have a good answer for. I agree that this is much simpler.

**Comment by quantumjot** [comment]
Changes made! Thanks for the great suggestions @jni - I'm pleased with how it's coming together.

**Review by jni — APPROVED** [review]
:tada:

**Comment by kevinyamauchi** [comment]
Awesome! Thanks, @quantumjot for this awesome application note! Also, thanks to @jni for the thorough reviews.

**Pull request merged**